### PR TITLE
Runner service implementation feature parity

### PR DESCRIPTION
**Pull request by mxsdev**
- [x] Run `background=true` as background tasks
- [x] User-prompts (UI) for ENV variables
- [x] Unify cmds/script prep?
- [x] Allow identification of grpc-created terminals

Needs changes in runme service:

- [x] Curl-ing img (last cell in `examples/README.md`) with notebook-native png renderer crash-hangs [#143](https://github.com/stateful/runme/issues/143)
- [x] Cat-ing LICENSE (in `examples/README.md`) does not work ~~[#144](https://github.com/stateful/runme/issues/144)~~ [#155](https://github.com/stateful/runme/issues/155)

### CODE REVIEW

File: src/extension/executors/runner.ts [hunk 1]
```diff
@@ -10,22 +10,25 @@ import {
...
 import { closeTerminalByEnvID } from './task'
-import { getShellPath } from './utils'
+import { getCommandExportExtractMatches, getShellPath, promptUserForVariable } from './utils'
 
 const LABEL_LIMIT = 15
+const BACKGROUND_TASK_HIDE_TIMEOUT = 2000
 const MIME_TYPES_WITH_CUSTOM_RENDERERS = ['text/plain']
 
 export async function executeRunner(
+  context: ExtensionContext,
```
Comment (sourishkrout) [inline src/extension/executors/runner.ts:31]: Side note: one of the reasons why I think this the function-abstraction of Executors won't suffice long-term. Not part of this PR, though.

File: src/extension/executors/runner.ts [hunk 2]
```diff
@@ -46,11 +88,9 @@ export async function executeRunner(
     programName: getShellPath(execKey),
     environment,
     script: {
```
Comment (sourishkrout) [inline src/extension/executors/runner.ts:90]: Consider renaming `script` to `exec` / `executable` or something like that. Script appears contradictory to commands being the name of one of many types.

File: src/extension/executors/utils.ts [hunk 3]
```diff
@@ -46,21 +98,18 @@ export async function retrieveShellCommand (exec: NotebookCellExecution) {
...
-    if (placeHolder.startsWith('$(') && placeHolder.endsWith(')')) {
+
+  for (const { hasStringValue, key, match, type, value } of exportMatches) {
+    if (type === 'exec') {
       /**
        * evaluate expression
        */
-      const expression = placeHolder.slice(2, -1)
-      const expressionProcess = cp.spawn(expression, {
+      const expressionProcess = cp.spawn(value, {
```
Comment (sourishkrout) [inline src/extension/executors/utils.ts:112]: Shouldn't this use the Runner APIs versus spawning child processes? Consider addressing in follow up issue/PR.

File: src/extension/utils.ts [hunk 4]
```diff
@@ -106,8 +111,8 @@ export function getKey(runningCell: vscode.TextDocument): keyof typeof executor
  * treat cells like like a series of individual commands
  * which need to be executed in sequence
  */
-export function getCmdShellSeq(cellText: string, os: string): string {
-  const trimmed = cellText
+export function getCmdSeq(cellText: string): string[] {
```
Comment (sourishkrout) [inline src/extension/utils.ts:114]: Once the TUI is using the Runner to execute commands we should consider GRPC exposing APIs to prep an input as a sequence of commands so both Ext/TUI plausible work the same way. Let's make a ticket for this if we haven't already.

**Review by sourishkrout — CHANGES_REQUESTED** [review]
Left some comments in the code. From a functional point of view, it appears that "export-only" cells have some unindented behavior.

Please see this clip: https://i.imgur.com/jqYi1yH.mp4

File: src/extension/executors/utils.ts (hunk 3 above)
Comment (mxsdev) [reply src/extension/executors/utils.ts:112]: This feature uses `getCommandExportExtractMatches` directly - it does not call `retrieveShellCommand`.

Comment (sourishkrout) [reply src/extension/executors/utils.ts:112]: great 👌 , thanks for clarifying!

**Comment by mxsdev** [comment]
@sourishkrout

Tested with the main example readme and every script worked for me. 

Could you test against really quick with the latest changes? Thanks!

File: src/extension/utils.ts (hunk 4 above)
Comment (mxsdev) [reply src/extension/utils.ts:114]: https://github.com/stateful/runme/issues/157

**Comment by sourishkrout** [comment]
> @sourishkrout
> 
> Tested with the main example readme and every script worked for me.
> 
> Could you test against really quick with the latest changes? Thanks!

Confirmed. Great job 👏 

Another thing I noticed is that `closeTerminalOnSuccess` does not seem to have an effect? Just on my end, @mxsdev?

**Review by sourishkrout — APPROVED** [review]
👍 LGTM

**Comment by mxsdev** [comment]
I was able to fix tests, but for the following line in `tests/extension/executors/utils.test.ts`:

```ts
vi.mock('../../../src/extension/utils', () => ({
  replaceOutput: vi.fn(),
}))
```

I was not able to use `importActual` as it would result in the process hanging. I guess this is due to cyclic dependency, but that's [supposed to be fixed](https://github.com/vitest-dev/vitest/issues/2567), so idk.

In the meantime I just mocked the method, but this is less than ideal since there's no need to mock it. Lmk if you think this is worth an issue @sourishkrout.

Anyways, this is still good to merge! 🎉
